### ISSUE TITLE
Consumer performance problems in 1.6.0

### ISSUE THREAD

**Issue by dimexid** [confluentinc/confluent-kafka-python#1054]
Description
===========
In our environment, we noticed significant performance change between the
- `1.6.0` version (version() ('1.6.0', 17170432), libversion() ('1.6.0', 17170687)) and
- `1.5.0` version (version() ('1.5.0', 17104896), libversion() ('1.5.0', 17105151))
by our **consumers** running on Debian Buster.

The older version manages to process all messages and the gap overall remains close to zero. If the newer version is installed in the same environment, the consumers performance drops significantly and the gap begins to grow.

Our client config looks like this:
```python
  "kafka_client_config": {
      "bootstrap.servers": "s1,s2,s3,s4,s5,s6,s7",
      "group.id": "group_id",
      "auto.offset.reset": "latest",
      "session.timeout.ms": 16000, 
      "heartbeat.interval.ms": 3900,
      "partition.assignment.strategy": "roundrobin,range",
  }
 ```
and we instantiate the client as follows:
```python
from confluent_kafka import Consumer as KafkaConsumer

...
consumer = KafkaConsumer(**kafka_client_config)
consumer.subscribe(topics)
...
```

In [librdkafka release notes](https://github.com/edenhill/librdkafka/releases/tag/v1.6.0) I noticed some notes on consumer changes:
>The consumer assignment and consumer group implementations have been
decoupled, simplified and made more strict and robust

Maybe the robustness could have some impact on the performance? Has somebody else noticed similar problem? Perhaps, we only missed some configuration change needed to use the newer `1.6.0` version..

Checklist
=========
Please provide the following information:

 - [x] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`):
 - [ ] Apache Kafka broker version:
 - [x] Client configuration: `{...}`
 - [x] Operating system:
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Do you have any comparative numbers to share? Preferably with processing commented out.

**Comment by dimexid** [comment]
We temporarily downgraded to 1.5.0 so I do not have any online stats. Leastwise I enclose relevant part of the dashboard monitoring our consumers.

![consumer_dashboard](https://user-images.githubusercontent.com/41079148/110793343-cce04880-8274-11eb-89c4-dcb3030d5219.png)

**Comment by edenhill** [comment]
I'm sorry but this is not enough information to work on.

**Comment by dimexid** [comment]
Which statistics would help you? I can try to find something in logs..

**Comment by edenhill** [comment]
Message consumption rate (preferably without processing): i.e., at what rate you are getting messages from poll() for both 1.5.x and 1.6.0.

**Comment by dimexid** [comment]
We use `consumer.consume(1000, 10)`. I tried both 1.5.x and 1.6.0 without processing and the consumption rate was perfectly comparable (~6900 msg/s).

**Comment by edenhill** [comment]
So you get the same raw message throughput on 1.5.x and 1.6.0, but with processing there's a throughput drop on 1.6.0 but not 1.5.x?

**Comment by dimexid** [comment]
Yes, it looks like so.. The processing I commented out includes e.g. `consumer.assignment()` or `consumer.get_watermark_offsets(topic_partition)` calls. I will try to benchmark them for both 1.5.x and 1.6.0.

**Comment by dimexid** [comment]
Finally, I reproduced the problem: If I run the consumer in multiple processes (I tried 10 and all the processes share the same `gropu.id`):
- `confluent-kafka==1.5.0`: all 10 processes are processing messages and every process has some partition assigned
- same setup, only `confluent-kafka==1.6.0`: only 3 of 10 processes are processing messages and 7 processes have no assigned partitions

The coverage is 100% in both cases but for 1.6.0, 7 processes do nothing..

**Comment by icsahn-rollbar** [comment]
we are seeing a similar behavior as well where just toggling between the version changes the performance of our Kafka consumer.

Our use case: we are trying to get a watermark offset for each partition of a topic. We are running 20 processes that read the watermark. When we do the watermark measurement with 
1. confluent-kafka==1.4.0 : Our latency remains around ~80ms
2. confluent-kafka==1.6.0 : Our latency increases to ~220ms

Mind that nothing else has changed except the version. Would be happy to provide more details if it helps.

**Comment by MSeal** [comment]
Closing for now given the lack of engagement. We're doing a performance pass in the near future to look at more recent version behavior herein.
